### PR TITLE
Removed code deprecated in qiskit-terra 0.21, released on June 2022

### LINKED ISSUE #10748: Remove code deprecated in 0.21 (released on June 30, 2022)
There is code released 13 month ago that needs to be removed.
 - qiskit/execute_function.py:39 (`execute`)
 - qiskit/execute_function.py:40 (`execute`)
 - qiskit/transpiler/passes/scheduling/alap.py:30 (`__init__`)
 - qiskit/transpiler/passes/scheduling/asap.py:36 (`__init__`)
 - qiskit/transpiler/passes/scheduling/dynamical_decoupling.py:93 (`__init__`)
 - qiskit/transpiler/passes/scheduling/alignments/align_measures.py:93 (`__init__`)
 - qiskit/transpiler/passes/utils/cx_direction.py:22 (`__init__`)
 - qiskit/transpiler/passes/utils/check_cx_direction.py:22 (`__init__`)
 - ~qiskit/algorithms/optimizers/spsa.py:652 (`optimize`)~ Not needed, as `qiskit.algorithms` has its own process.
 - qiskit/pulse/transforms/alignments.py:48 (`to_dict`)
 - qiskit/pulse/transforms/alignments.py:333 (`to_dict`)
 - qiskit/pulse/transforms/alignments.py:419 (`to_dict`)
 - qiskit/qpy/interface.py:75 (`dump`)
  
This includes:
- [ ] Remove deprecated code and tests (if they only check the deprecation raise)
- [ ] tag the PR as [`Changelog: Removal`](https://github.com/Qiskit/qiskit/labels/Changelog%3A%20Removal)
- [ ] create a release note in the category `upgrade:`. If you can include an example with an alternative for user to migrate to the new code, as this change might break users code.

### PR DESCRIPTION
**Pull request by Raghav-Bell**
It fixes #10748 

### Details and comments

~Please tag it as [`Changelog: Removal`](https://github.com/Qiskit/qiskit/labels/Changelog%3A%20Removal)~

### CODE REVIEW

**Comment by Raghav-Bell** [comment]
`tox` test fails on `test\benchmarks\scheduling_passes.py` as it  only contains check for `TimeUnitConversion` . Should we keep `test\benchmarks\scheduling_passes.py` or not ? also i have removed most of the  tests for depreciated or removed `classes`. Please let me know if i am missing something.

**Comment by woodsp-ibm** [comment]
I am unsure about `optimizers/spsa` - mainly in regards of whatever upgrade note you do for the release note (which has yet to be included here). The optimizers are part of algorithms - I did have a prior issue #9866 that included addressing spsa but when the whole of algorithms was deprecated in the last release I simply closed it. I know in other similar issues that @1ucian0 created any deprecation removal for algorithms was dropped - maybe that's the simplest course here @1ucian0? - not to change spsa here given the whole module its part of is already deprecated - that will avoid any issues around including it or not in the release note

**Comment by ElePT** [comment]
I'm with @woodsp-ibm, I think it's best to leave algorithms-related code untouched until its removal.

**Comment by 1ucian0** [comment]
I agree that leaving out `qiskit.algorithms` from this removal is probably the best option (the original list was created by a program).

My apologies for including it originally.

File: qiskit/algorithms/optimizers/spsa.py [hunk 1]
```diff
@@ -666,14 +666,12 @@ def optimize(
         initial_point=None,
     ):
         """Perform optimization.
-
         Args:
```
Comment (1ucian0) [inline qiskit/algorithms/optimizers/spsa.py]:
```suggestion

        Args:
```

File: qiskit/algorithms/optimizers/spsa.py [hunk 2]
```diff
@@ -666,14 +666,12 @@ def optimize(
...
             num_vars (int): Number of parameters to be optimized.
             objective_function (callable): A function that computes the objective function.
             gradient_function (callable): Not supported for SPSA.
             variable_bounds (list[(float, float)]): Not supported for SPSA.
             initial_point (numpy.ndarray[float]): Initial point.
-
         Returns:
```
Comment (1ucian0) [inline qiskit/algorithms/optimizers/spsa.py]:
```suggestion

        Returns:
```

**Review by 1ucian0 — CHANGES_REQUESTED** [review]
I moving this PR as draft, as it seems to need some extra work before reviewing. Don't forget to add a release note:
> create a release note in the category upgrade:. If you can include an example with an alternative for user to migrate to the new code, as this change might break users code.

**Comment by 1ucian0** [comment]
Take a look to the conflicts! When ready for review, hit `Ready for review`.

**Comment by Raghav-Bell** [comment]
@1ucian0 These conflicts are due to  recent merge of PR #10753  as in the PR #10753  `classes ALAPSchedule, AlignMeasures, DynamicalDecoupling` are changed which are deprecated in `0.21` (removed with merge of this PR).  I think we can merge this PR. Please let me know what do you think on this ?

**Comment by 1ucian0** [comment]
> These conflicts are due to recent merge of PR #10753 as in the [...] Please let me know what do you think on this ?

I mean, by "take a look" I wanted to say "resolve the conflicts". Sometimes my word choice is not the best. My bad.

**Comment by Raghav-Bell** [comment]
@1ucian0 It seems these conflicts can not be resolved by me (refer: [How do I fix a merge conflict due to removal of a file in a branch?](https://stackoverflow.com/questions/1380670/how-do-i-fix-a-merge-conflict-due-to-removal-of-a-file-in-a-branch)) Maintainers should remove `qiskit/transpiler/passes/scheduling/alap.py`, `qiskit/transpiler/passes/scheduling/alignments/align_measures.py` and `qiskit/transpiler/passes/scheduling/dynamical_decoupling.py` while merging .  If that is not the case please help me to resolve the conflicts .

**Edit:**
I experimented and fixed the conflicts . This PR helped me learning a new fix , Thanks @1ucian0 for assigning me this .

**Comment by Raghav-Bell** [comment]
Following test is failing on windows , let me check the fix and resolve the conflict:
https://github.com/Qiskit/qiskit/blob/e62c86bbc2720883270641f1c154743a657cfca8/test/python/algorithms/test_measure_error_mitigation.py#L56

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    :class:`.transpiler.passes.DynamicalDecoupling` is removed instead
+    use :class:`.transpiler.passes.PadDynamicalDecoupling`.
+        
+  - |
+    :class:`.transpiler.passes.AlignMeasures` is removed instead
+    use :class:`.transpiler.passes.ConstrainedReschedule`.
+
+  - |
+    :class:`.transpiler.passes.CXDirection` is removed instead
+    use generic :class:`.transpiler.passes.GateDirection`.
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
    The transpilation pass ``qiskit.transpiler.passes.CXDirection`` is removed. It use was deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
    Instead, use the more generic :class:`~.GateDirection` pass.
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
+---
+upgrade:
+  - |
+    Deprecated code and tests in 0.21 are removed.
+    :func:`~qiskit.execute_function.execute` doesn't support
+    `qobj_id` and `qobj_header` arguments.
+
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
    The function :func:`~qiskit.execute_function.execute`
    does not accept the arguments `qobj_id` and  `qobj_header` any more.
    Their use was deprecated in Qiskit 0.37 (with Terra 0.21), released on June 2022.    
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+upgrade:
+  - |
+    Deprecated code and tests in 0.21 are removed.
+    :func:`~qiskit.execute_function.execute` doesn't support
+    `qobj_id` and `qobj_header` arguments.
+    
+  - |
+    :class:`.transpiler.passes.ALAPSchedule` is removed instead
+    use :class:`.transpiler.passes.ALAPScheduleAnalysis`.
+
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
    The transpilation pass ``qiskit.transpiler.passes.ALAPSchedule`` is removed. It use was deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022 and replaced by 
    :class:`~.transpiler.passes.ALAPScheduleAnalysis`, which is an
    analysis pass that require
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    `qobj_id` and `qobj_header` arguments.
+    
+  - |
+    :class:`.transpiler.passes.ALAPSchedule` is removed instead
+    use :class:`.transpiler.passes.ALAPScheduleAnalysis`.
+    
+  - |
+    :class:`.transpiler.passes.ASAPSchedule` is removed instead
+    use :class:`.transpiler.passes.ASAPScheduleAnalysis`.
+
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
  - |
    The transpilation pass ``qiskit.transpiler.passes.ASAPSchedule`` is removed. It use was deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022. It has been superseded by
    :class:`~.ASAPScheduleAnalysis` and the new scheduling workflow.
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    use :class:`.transpiler.passes.ALAPScheduleAnalysis`.
+    
+  - |
+    :class:`.transpiler.passes.ASAPSchedule` is removed instead
+    use :class:`.transpiler.passes.ASAPScheduleAnalysis`.
+    
+  - |
+    :class:`.transpiler.passes.DynamicalDecoupling` is removed instead
+    use :class:`.transpiler.passes.PadDynamicalDecoupling`.
+
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
    The transpilation pass ``qiskit.transpiler.passes.DynamicalDecoupling`` is removed. It use was deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
    Instead, use :class:`~.transpiler.passes.PadDynamicalDecoupling`, which performs the same
    function but requires scheduling and alignment analysis passes to run prior to it.
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    use :class:`.transpiler.passes.ASAPScheduleAnalysis`.
+    
+  - |
+    :class:`.transpiler.passes.DynamicalDecoupling` is removed instead
+    use :class:`.transpiler.passes.PadDynamicalDecoupling`.
+        
+  - |
+    :class:`.transpiler.passes.AlignMeasures` is removed instead
+    use :class:`.transpiler.passes.ConstrainedReschedule`.
+
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
    The transpilation pass ``qiskit.transpiler.passes.AlignMeasures`` is removed. It use was deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
    Instead, use :class:`~.ConstrainedReschedule`, which performs the same function 
    and also supports aligning to additional timing constraints.
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  - |
+    :class:`.transpiler.passes.CXDirection` is removed instead
+    use generic :class:`.transpiler.passes.GateDirection`.
+
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
    The transpilation pass ``qiskit.transpiler.passes.CXDirection`` is removed. It use was deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
    Instead, use the more generic :class:`~.GateDirection` pass.
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  - |
+    :class:`.transpiler.passes.CheckCXDirection` is removed instead
+    use generic :class:`.transpiler.passes.CheckGateDirection`.
+
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
    The transpilation pass ``qiskit.transpiler.passes.CheckCXDirection`` is removed. It use was deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
    Instead, use the more generic :class:`~.CheckGateDirection` pass.
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  - |
+    :class:`.pulse.transforms.AlignmentKind`, `.pulse.transforms.AlignEquispaced`,
+    and :class:`.pulse.transforms.AlignFunc` don't support :meth:`to_dict`.
+
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
    The methods ``to_dict`` in the classes :class:`.pulse.transforms.AlignmentKind`,
    `.pulse.transforms.AlignEquispaced`, and :class:`.pulse.transforms.AlignFunc` are removed.
    They were deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
```

Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml:44]:
```suggestion
- |
   The argument ``circuits`` in the method :meth:`qiskit.qpy.interface.dump`
   is removed as its usage was deprecated 
   in Qiskit 0.37 (with Terra 0.21), released on June 2022.
   Instead, use the argument ``programs``, which behaves identically.
```

File: test/benchmarks/scheduling_passes.py
Comment (1ucian0) [inline test/benchmarks/scheduling_passes.py:1]: I'm not fully sure, but I think these tests should be migrated, as they are `test/benchmarks`

File: test/python/transpiler/test_instruction_alignments.py
Comment (1ucian0) [inline test/python/transpiler/test_instruction_alignments.py:1]: Similarly, these tests should be migrated to the new workflow, I think.

**Review by 1ucian0 — CHANGES_REQUESTED** [review]
Thanks for taking care of this! It is a complicated one. I think some of the tests should be migrated instead of removed. Would you like to have a shoot at that?

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml (hunk 11 above)
Comment (Raghav-Bell) [reply releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml:44]: From next PR onwards, I will try to make release notes more user oriented. 😅

File: test/benchmarks/scheduling_passes.py
Comment (Raghav-Bell) [reply test/benchmarks/scheduling_passes.py:1]: I think they are already migrated here in [benchmarks/scheduling_passes.py](https://github.com/Qiskit/qiskit/blob/main/test/benchmarks/scheduling_passes.py).

File: test/python/transpiler/test_instruction_alignments.py
Comment (Raghav-Bell) [reply test/python/transpiler/test_instruction_alignments.py:1]: I have checked similiar tests in [benchmarks/transpiler_levels.py](https://github.com/Qiskit/qiskit/blob/main/test/benchmarks/transpiler_levels.py). Can you please elaborate about this? Should we add a github workflow for them ?

File: qiskit/transpiler/passes/__init__.py [hunk 12]
```diff
@@ -114,13 +114,10 @@
...
    PadDelay
    ConstrainedReschedule
-   AlignMeasures
    ValidatePulseGates
    InstructionDurationCheck
    SetIOLatency
-   ALAPSchedule
-   ASAPSchedule
-   DynamicalDecoupling
+
```
Comment (1ucian0) [inline qiskit/transpiler/passes/__init__.py]:
```suggestion
```

File: qiskit/execute_function.py [hunk 13]
```diff
@@ -279,8 +265,7 @@ def execute(
 
             job = execute(qc, backend, shots=4321)
     """
-    del qobj_id
-    del qobj_header
+
```
Comment (1ucian0) [inline qiskit/execute_function.py]:
```suggestion
```

File: qiskit/transpiler/passes/scheduling/__init__.py [hunk 14]
```diff
@@ -12,9 +12,7 @@
 
 """Module containing circuit scheduling passes."""
 
-from .alap import ALAPSchedule
-from .asap import ASAPSchedule
-from .dynamical_decoupling import DynamicalDecoupling
+
```
Comment (1ucian0) [inline qiskit/transpiler/passes/scheduling/__init__.py]:
```suggestion
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 15]
```diff
@@ -0,0 +1,52 @@
+---
+upgrade:
+  - |
+    The function :func:`~qiskit.execute_function.execute`
+    does not accept the arguments `qobj_id` and  `qobj_header` any more.
+    Their use was deprecated in Qiskit 0.37 (with Terra 0.21), released on June 2022.    
+
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    The function :func:`~qiskit.execute_function.execute`
+    does not accept the arguments `qobj_id` and  `qobj_header` any more.
+    Their use was deprecated in Qiskit 0.37 (with Terra 0.21), released on June 2022.    
+    
+  - |
+    The transpilation pass ``qiskit.transpiler.passes.ALAPSchedule`` is removed. It use was deprecated 
+    in Qiskit 0.37 (with Terra 0.21), released on June 2022 and replaced by 
+    :class:`~.transpiler.passes.ALAPScheduleAnalysis`, which is an
+    analysis pass.
+
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    The transpilation pass ``qiskit.transpiler.passes.ALAPSchedule`` is removed. It use was deprecated 
+    in Qiskit 0.37 (with Terra 0.21), released on June 2022 and replaced by 
+    :class:`~.transpiler.passes.ALAPScheduleAnalysis`, which is an
+    analysis pass.
+    
+  - |
+    The transpilation pass ``qiskit.transpiler.passes.ASAPSchedule`` is removed. It use was deprecated 
+    in Qiskit 0.37 (with Terra 0.21), released on June 2022. It has been superseded by
+    :class:`~.ASAPScheduleAnalysis` and the new scheduling workflow.
+
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
```

File: releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+    The methods ``to_dict`` in the classes :class:`.pulse.transforms.AlignmentKind`,
+    `.pulse.transforms.AlignEquispaced`, and :class:`.pulse.transforms.AlignFunc` are removed.
+    They were deprecated 
+    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
+
+  - |
+    The argument ``circuits`` in the method :meth:`qiskit.qpy.interface.dump`
+    is removed as its usage was deprecated 
+    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
+    Instead, use the argument ``programs``, which behaves identically.
```
Comment (1ucian0) [inline releasenotes/notes/removed_deprecated_0.21-6c93f7bbc50ae40e.yaml]:
```suggestion
  - |
    The transpilation pass ``qiskit.transpiler.passes.AlignMeasures`` is removed. It use was deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
    Instead, use :class:`~.ConstrainedReschedule`, which performs the same function 
    and also supports aligning to additional timing constraints.
  - |
    The transpilation pass ``qiskit.transpiler.passes.CXDirection`` is removed. It use was deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
    Instead, use the more generic :class:`~.GateDirection` pass.
  - |
    The transpilation pass ``qiskit.transpiler.passes.CheckCXDirection`` is removed. It use was deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
    Instead, use the more generic :class:`~.CheckGateDirection` pass.
  - |
    The methods ``to_dict`` in the classes :class:`.pulse.transforms.AlignmentKind`,
    `.pulse.transforms.AlignEquispaced`, and :class:`.pulse.transforms.AlignFunc` are removed.
    They were deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
  - |
    The argument ``circuits`` in the method :meth:`qiskit.qpy.interface.dump`
    is removed as its usage was deprecated 
    in Qiskit 0.37 (with Terra 0.21), released on June 2022.
    Instead, use the argument ``programs``, which behaves identically.
```

File: qiskit/execute_function.py [hunk 19]
```diff
@@ -26,7 +26,7 @@
 from qiskit.providers.backend import Backend
 from qiskit.pulse import Schedule, ScheduleBlock
 from qiskit.exceptions import QiskitError
-from qiskit.utils.deprecation import deprecate_arg
+
```
Comment (1ucian0) [inline qiskit/execute_function.py:29]:
```suggestion
```

**Review by 1ucian0 — APPROVED** [review]
I migrated the benchmark tests to the new style. The rest, let's merge. Thanks!

**Pull request merged**